### PR TITLE
[FLINK-22725][coordination] SlotManagers unregister metrics in suspend()

### PR DESCRIPTION
**Pull request by zentol**
Modified version of #15904.

Slotmanagers register metrics in start(), but only unregister them in close().

This has 2 issues:

a) If the SM is restarted it cannot re-register the metrics because the old ones are still present; this isn't a critical problem (because the old ones still work) but it produces unnecessary logging noise

b) The metric may produce an NPE when accessed by a reporter during suspend().

### CODE REVIEW

**Comment by fapaul** [comment]
Why does this issue only affect the `FineGrainedSlotManager` and `DeclarativeSlotMager` when looking at the `SlotManagerImpl` it also calls `suspend` before unregistering the metrics?

**Comment by zentol** [comment]
The SlotManagerImpl will be removed in 1.14, so there's no value in patching that.
But I'll keep it in mind for the 1.13 backport.

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManagerTest.java [hunk 1]
```diff
@@ -948,4 +954,41 @@ private void testMaxTotalResourceExceededRegisterResource(
...
+        final Context context = new Context();
+        context.setSlotManagerMetricGroup(
+                SlotManagerMetricGroup.create(metricRegistry, "localhost"));
+
+        context.runTest(
+                () -> {
+                    context.runInMainThreadAndWait(
+                            () -> {
+                                try {
+                                    closeFn.accept(context.getSlotManager());
```
Comment (fapaul) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManagerTest.java:986]: Nit: Assert that the number of registered metrics is greater than 0 before calling the closing method.

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManagerTest.java [hunk 2]
```diff
@@ -1450,6 +1454,34 @@ public void testClearRequirementsClearsResourceTracker() throws Exception {
...
+                TestingMetricRegistry.builder()
+                        .setRegisterConsumer((a, b, c) -> registeredMetrics.incrementAndGet())
+                        .setUnregisterConsumer((a, b, c) -> registeredMetrics.decrementAndGet())
+                        .build();
+
+        final DeclarativeSlotManager slotManager =
+                createDeclarativeSlotManagerBuilder()
+                        .setSlotManagerMetricGroup(
+                                SlotManagerMetricGroup.create(metricRegistry, "localhost"))
+                        .buildAndStartWithDirectExec();
```
Comment (fapaul) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManagerTest.java:1480]: Nit: Assert that metrics number is greater than 0.

**Review by fapaul — APPROVED** [review]
Thanks for the fix. The change looks good I just left two minor comments to perhaps harden the test cases.

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManagerTest.java [hunk 3]
```diff
@@ -948,4 +955,42 @@ private void testMaxTotalResourceExceededRegisterResource(
...
+                        context.runInMainThreadAndWait(
+                                () -> {
+                                    try {
+                                        // sanity check to ensure metrics were actually registered
+                                        assertThat(registeredMetrics.get(), greaterThan(0));
+                                        closeFn.accept(context.getSlotManager());
+                                        assertThat(registeredMetrics.get(), is(0));
+                                    } catch (Exception e) {
+                                        fail("Error when closing slot manager.");
+                                    }
```
Comment (KarmaGYZ) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManagerTest.java:993]: Not sure is it a good way to do the sanity check in the `context.runInMainThreadAndWait`. The error from `assertThat` will not be logged.

I think we can leverage a `CompleteFuture` to check whether the close function is done successfully and do the sanity check out of the main thread. Also, we do not need the `runInMainThreadAndWait` in that case.

**Review by KarmaGYZ — APPROVED** [review]
LGTM.

**Pull request merged**